File: 52_get_intersection_node/get_intersection_node.go
/*
 * @Date: 2021-07-21 18:18:20
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-07-21 19:47:34
 */

package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func getIntersectionNode(headA, headB *ListNode) *ListNode {
	if headA == nil || headB == nil {
		return nil
	}
	pa, pb := headA, headB
	for pa != pb {
		if pa == nil {
			pa = headB
		} else {
			pa = pa.Next
		}
		if pb == nil {
			pb = headA
		} else {
			pb = pb.Next
		}
	}
	return pa
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	{
		intersection := &ListNode{8, &ListNode{4, &ListNode{5, nil}}}
		headA := &ListNode{4, &ListNode{1, intersection}}
		headB := &ListNode{5, &ListNode{0, &ListNode{1, intersection}}}
		assert(getIntersectionNode(headA, headB).Val == 8)
	}
	{
		intersection := &ListNode{2, &ListNode{4, nil}}
		headA := &ListNode{3, intersection}
		headB := &ListNode{0, &ListNode{9, &ListNode{1, intersection}}}
		assert(getIntersectionNode(headA, headB).Val == 2)
	}
	{
		headA := &ListNode{1, &ListNode{5, nil}}
		headB := &ListNode{2, &ListNode{6, &ListNode{4, nil}}}
		assert(getIntersectionNode(headA, headB) == nil)
	}
}

File: 52_get_intersection_node/get_intersection_node.cpp
/*
 * @Date: 2021-07-21 18:18:15
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-07-21 18:26:04
 */

#include <cassert>

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
 public:
  ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
    if (headA == nullptr || headB == nullptr) {
      return nullptr;
    }
    ListNode *pA = headA, *pB = headB;
    while (pA != pB) {
      pA = pA == nullptr ? headB : pA->next;
      pB = pB == nullptr ? headA : pB->next;
    }
    return pA;
  }
};

int main() {
  {
    ListNode *intersection =
        new ListNode{8, new ListNode{4, new ListNode{5, nullptr}}};
    ListNode *headA = new ListNode{4, new ListNode{1, intersection}};
    ListNode *headB =
        new ListNode{5, new ListNode{0, new ListNode{1, intersection}}};
    assert(Solution{}.getIntersectionNode(headA, headB)->val == 8);
  }
  {
    ListNode *intersection =
        new ListNode{2, new ListNode{4, nullptr}};
    ListNode *headA = new ListNode{0, new ListNode{9, new ListNode{1, intersection}}};
    ListNode *headB =
        new ListNode{3,  intersection};
    assert(Solution{}.getIntersectionNode(headA, headB)->val == 2);
  }
  {
    ListNode *headA = new ListNode{2, new ListNode{6, new ListNode{4, nullptr}}};
    ListNode *headB =
        new ListNode{1,  new ListNode{5, nullptr}};
    assert(Solution{}.getIntersectionNode(headA, headB) == nullptr);
  }
}